[14.0] [IMP] cetmix_tower_server: managers can remove flightplan lines

Aldeigja opened the pull request:
### Before this PR

- **In plan lines**
  - Members of the group_manager cannot delete the plan lines they have created
  - Members of the manager and user groups can access plan lines assigned
     to plans belonging to servers to which these users are not subscribed.

- **In plan line actions**
  - Members of the group_manager cannot delete the plan line actions they have created.
  - Members of the manager and user groups can access plan line actions assigne
     to plans belonging to servers to which these users are not subscribed.

- **In variables assigned to plan line actions**
   -  members of group_user and group_manager have no access to variable values related to plan line actions

- **In cx.tower.access.mixin**
  - 'access_level' field was restricted to 'Manager' and 'Root' groups only.
  It caused access error when member 'User' group tried to access it.
  Although it appeared only in tests it could have created potential
  issues in other cases.

- **In variables assigned to server templates**
  - Managers can remove variable value only if they subscribed to server template

### After this PR

- **In plan lines**
  - Members of the group_manager can delete the plan lines they have created if
    plans which these assigned doesn't belong to any server or belong to server
    to which this user subscribed
  - Members of the manager and user groups can not access plan lines assigned
    to plans belonging to servers to which these users are not subscribed.

- **In plan line actions**
   - Members of the group_manager can delete the plan line actions they have created if
     plans which these assigned doesn't belong to any server or belong to server
     to which this user subscribed
   - Members of the manager and user groups can not access plan line actions assigned
     to plans belonging to servers to which these users are not subscribed

- **In variables assigned to plan line actions**
If they are subscribed to the servers associated to plans parent for these actions, then:
    - members of group user can read variable values related to plan line actions
    - members of group manager can:
      -  read variable values related to plan line actions
      -  create variable values related to plan line actions
      -  delete variable values related to plan line actions  they have created

- **In cx.tower.access.mixin**
  - Remove group restriction from 'access_level' field.
    If needed they can be applied at the view level.

Task #3510
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Security Enhancements**
	- Updated access control rules for users and managers across multiple system components.
	- Refined permissions for plan lines, plan line actions, and variable values, including new unlink permissions for managers.
	- Introduced distinct access rules for user and manager roles, enhancing granularity in permissions.

- **Testing**
	- Expanded test coverage for access rights, focusing on user and manager permissions.
	- Introduced new tests to verify access control mechanisms for plan lines, plan line actions, and variable values in server templates.

- **Code Improvements**
	- Cleaned up security XML configurations for better readability.
	- Enhanced permission checks for different user groups.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Aldeigja committed:
[IMP] cetmix_tower_server: managers can delete own plan line actions

        Before this commit
        ------------------
        - Members of the group_manager cannot delete the plan line actions they have created.
        - Members of the manager and user groups can access plan line actions assigned
	 to plans belonging to servers to which these users are not subscribed.

        After this commit
        ------------------
       - Members of the group_manager can delete the plan line actions they have created if
         plans which these assigned doesn't belong to any server or belong to server
         to which this user subscribed
       - Members of the manager and user groups can not access plan line actions assigned
         to plans belonging to servers to which these users are not subscribed

ivs-cetmix commented:
/ocabot rebase

CetmixGitDrone commented:
Congratulations, PR rebased to [14.0-dev](https://github.com/cetmix/cetmix-tower/commits/14.0-dev).

Commit:
[IMP] cetmix_tower_server: managers can delete own plan lines
Before this commit
------------------

- Members of the group_manager cannot delete the plan lines they have created
- Members of the manager and user groups can access plan lines assigned
  to plans belonging to servers to which these users are not subscribed.

After this commit
------------------
- Members of the group_manager can delete the plan lines they have created if
  plans which these assigned doesn't belong to any server or belong to server
  to which this user subscribed
- Members of the manager and user groups can not access plan lines assigned
  to plans belonging to servers to which these users are not subscribed.

Aldeigja committed:
[IMP] cetmix_tower_server: update variable values access rules:
Before this commit:
memnbers of group_user and group_manager have no access to variable values related to plan line actions

After this commit:
members of group user can read variable values related to plan line actions
members of group manager can
-  read variable values related to plan line actions
-  delete variable values related to plan line actions  they have created

Commit:
[FIX] cetmix_tower_server Access mixin

cx.tower.access.mixin

Before this commit
------------------

'access_level' field was restricted to 'Manager' and 'Root' groups only.
It caused access error when member 'User' group tried to access it.
Although it appeared only in tests it could have created potential
issues in other cases.

After this commit
-----------------

Remove group restriction from 'access_level' field.
If needed they can be applied at the view level.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -13,6 +17,59 @@
+        <field name="model_id" ref="model_cx_tower_variable_value" />
+        <field name="domain_force">
+            [
+            ('plan_line_action_id.access_level', '=', '1')
```
What if access level == 1, but this plan is linked to specific servers and current used doesn't have access to them?

Aldeigja replied:
I've updated the rules

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -13,6 +17,59 @@
+        <field name="model_id" ref="model_cx_tower_variable_value" />
+        <field name="domain_force">
+            [
+            ('plan_line_action_id.access_level', '&lt;=', '2')
```
What if access level <= 2, but this plan is linked to specific servers and current used doesn't have access to them?

Aldeigja committed:
[IMP] cetmix_tower_server: update variable values access rules in pline line actions:
Before this commit
------------------

    group_user and group_manager users have access to the variable values  associated
    with the plan line actions regardless of whether or not they were or were not
    subscribed to the servers associated with the parent plans for those plan line actions

After this commi
---------------

If they are subscribed to the servers associated to plans parent for these actions, then:

    members of group user can read variable values related to plan line actions
    members of group manager can
    -  read snd update variable values related to plan line actions
    -  delete variable values they have created

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_plan_line_action_security.xml:
```diff
@@ -6,18 +6,54 @@
+            '|',
+            ('line_id.plan_id.server_ids', '=', False),
+            ('line_id.plan_id.server_ids.message_partner_ids', 'in', [user.partner_id.id])
+
```
Please remove the blank line

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_plan_line_action_security.xml:
```diff
@@ -6,18 +6,54 @@
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_manager'))]" />
+        <field name="perm_unlink" eval="1" />
```
"1" is the default value for the "perm_..." fields. Which means you need to set it to "0" explicitly. Otherwise this rule is applied to all other actions too.
NB: this case should have been caught by tests, looks like they should be fixed or updated.

Aldeigja replied:
Thank you for noticing, you are right, as in this case manager should have full access to the plan_line_action he has created, I don't need set this value to "1" expicitly, as this is default.

Test cases related to the manager's ability to delete plan_line_actions are covered [here](https://github.com/cetmix/cetmix-tower/blob/75c5fb668cf66618af4bd5886f43451fdfac5c85/cetmix_tower_server/tests/test_plan.py#L1483) and below, if you think it is insufficient please let me know.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_plan_line_security.xml:
```diff
@@ -6,18 +6,51 @@
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_manager'))]" />
+        <field name="perm_unlink" eval="1" />
```
"1" is the default value for the "perm_..." fields. Which means you need to set it to "0" explicitly. Otherwise this rule is applied to all other actions too.
NB: this case should have been caught by tests, looks like they should be fixed or updated.

Aldeigja replied:
Thank you for noticing, you are right, as in this case manager should have full access to the plan_line he has created, I don't need set this value to "1" expicitly, as this is default.

Test cases related to the manager's ability to delete plan_line are covered [here](https://github.com/cetmix/cetmix-tower/blob/75c5fb668cf66618af4bd5886f43451fdfac5c85/cetmix_tower_server/tests/test_plan.py#L1321) and below, if you think it isn't enough please let me know.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,16 +1,82 @@
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_manager'))]" />
+        <field name="perm_unlink" eval="1" />
```
"1" is the default value for the "perm_..." fields. Which means you need to set it to "0" explicitly. Otherwise this rule is applied to all other actions too.
NB: this case should have been caught by tests, looks like they should be fixed or updated.

Aldeigja replied:
fixed

Aldeigja commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,16 +1,82 @@
         />
+    </record>
+
+    <record id="cx_tower_variable_value_action_rule_group_user_access" model="ir.rule">
+        <field
+            name="name"
+        >Tower variable value: user access to variable values in plan line action
+            rule</field>
+        <field name="model_id" ref="model_cx_tower_variable_value" />
+        <field name="domain_force">
+            [
+            ('plan_line_action_id.access_level', '=', '1'),
+            '|',
+            ('plan_line_action_id.line_id.plan_id.server_ids', '=', False),
+            ('plan_line_action_id.line_id.plan_id.server_ids.message_partner_ids', 'in', [user.partner_id.id])
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_user'))]" />
+    </record>
```
This is not relevant comment

Aldeigja commented on cetmix_tower_server/security/cx_tower_plan_line_security.xml:
```diff
@@ -6,18 +6,48 @@
+    </record>
 
 
+    <record id="cx_tower_plan_line_rule_group_manager_unlink" model="ir.rule">
+        <field name="name">Tower plan line: manager delete own records</field>
+        <field name="model_id" ref="model_cx_tower_plan_line" />
+        <field name="domain_force">
+            [
+            ('create_uid', '=', user.id),
+            ('access_level', '&lt;=', '2'),
+            '|',
+            ('plan_id.server_ids', '=', False),
+            ('plan_id.server_ids.message_partner_ids', 'in', [user.partner_id.id])
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_manager'))]" />
```
In case defined by domain in this rule, user should have full access to the 
model_cx_tower_plan_line record, so why do you suggest setting the rules this way:

 <field name="perm_unlink" eval="1"/>  
        <field name="perm_read" eval="0"/>  
        <field name="perm_write" eval="0"/>  
        <field name="perm_create" eval="0"/>  

Wouldn't it be better to just remove all perm_ fields in this rule because we need to use default values?

Aldeigja committed:
[IMP] cetmix_tower_server: update variable values access rules in templates:
    Before this commit
    ------------------

        members of group_manager don't have access to the variable values associated
        with the server templates.

    After this commit
    ---------------

    If members of group manager  are subscribed to the server template then:
    members of group manager can
        -  read and update variable values related to server templates they are subscribed
        -  delete variable values they have created

Aldeigja commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,115 @@
         />
+    </record>
+
+    <record id="cx_tower_variable_value_action_rule_group_user_access" model="ir.rule">
+        <field
+            name="name"
+        >Tower variable value: user access to variable values in plan line action
+            rule</field>
+        <field name="model_id" ref="model_cx_tower_variable_value" />
+        <field name="domain_force">
+            [
+            ('plan_line_action_id.access_level', '=', '1'),
+            '|',
+            ('plan_line_action_id.line_id.plan_id.server_ids', '=', False),
+            ('plan_line_action_id.line_id.plan_id.server_ids.message_partner_ids', 'in',
+            [user.partner_id.id])
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_user'))]" />
+    </record>
```
access_level is the field defined in the [cx_tower_access_mixin ](https://github.com/cetmix/cetmix-tower/blob/fd6edf728b21e06108c238ddd58b3900e6474a29/cetmix_tower_server/models/cx_tower_access_mixin.py#L35) it has nothing to do with perm_read etc., so your comment not applicable here.

ivs-cetmix commented on cetmix_tower_server/security/ir.model.access.csv:
```diff
@@ -14,7 +14,7 @@ access_server_user,Server->User,model_cx_tower_server,group_user,1,0,0,0
 access_server_root,Server->Root,model_cx_tower_server,group_root,1,1,1,1
 access_interpreter_user,Interpreter->User,model_cx_tower_interpreter,group_user,1,0,0,0
-access_interpreter_manager,Interpreter->Manager,model_cx_tower_interpreter,group_manager,1,1,1,0
+access_interpreter_manager,Interpreter->Manager,model_cx_tower_interpreter,group_manager,1,1, 1,0
```
Please remove extra space here

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,114 @@
+            ('plan_line_action_id.line_id.plan_id.server_ids.message_partner_ids', 'in',
+            [user.partner_id.id])
+            ]
+        </field>
```
You should specify that this rule is applied to "unlink" action only.  To avoid unnecessary  checks while performing other operations.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,114 @@
     </record>
 
+    <record
+        id="cx_tower_variable_value_template_rule_group_manager_access"
```
You should specify that it's a  `..server_template..` here because there are other templates in the system.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,114 @@
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_manager'))]" />
+    </record>
```
You should specify that this rule is applied to "unlink" action only.  To avoid unnecessary  checks while performing other operations.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,114 @@
+    </record>
+
+    <record
+        id="cx_tower_variable_value_template_rule_group_manager_unlink"
```
You should specify that it's a  `..server_template..` here because there are other templates in the system.

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,425 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        )
+
+        # Check unlink failure due to follower dependency
+        # Create a new plan line action as user_bob
```
Probably you can do this test before the previous one. So you can simply use existing action line instead of creating a new one again.

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,425 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        variable_value = self.env["cx.tower.variable.value"].create(
+            {
+                "variable_id": self.variable_os.id,
+                "value_char": "Branch = Main",
```
Value looks weird and may be misleading.

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,425 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        # Plan not assigned to server, ensure access to variable value in plan line
+        # action
+        # Add user to group
+        self.add_to_group(self.user_bob, "cetmix_tower_server.group_user")
```
You have removed this user from group in line 1582. Why are you adding it back again here? Maybe simply not to remove it at all?

Aldeigja replied:
Thanks for the comment, 
I removed user_bob in line 1582 from all groups, because this user is used in a large number of tests throughout the module, so I'm not sure if he doesn't belong to any tower.server group at the moment.
I encountered a similar situation a few months earlier.   So I did this just to be sure.
Perhaps there is another, more efficient way to do this check, but I don't know about it.

ivs-cetmix reviewed (changes requested):
Consider moving out some common methods into `setUp`.
Also you can remove some intermediate operations by optimising the test flow.

Aldeigja commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,120 @@
+        <field name="perm_unlink" eval="0" />
+    </record>
+
+    <record
+        id="cx_tower_variable_value_server_template_rule_group_manager_unlink"
+        model="ir.rule"
+    >
+        <field
+            name="name"
+        >Tower variable value: manager delete own variable values in server
+            template rule</field>
+        <field name="model_id" ref="model_cx_tower_variable_value" />
+        <field name="domain_force">
+            [
+            ('create_uid', '=', user.id)
+            ]
+        </field>
+        <field name="groups" eval="[(4, ref('cetmix_tower_server.group_manager'))]" />
+        <field name="perm_read" eval="0" />
+        <field name="perm_create" eval="0" />
+        <field name="perm_write" eval="0" />
+    </record>
```
> Group rules unify, if two group rules apply then either can be satisfied for the access to be granted. This means adding group rules can expand access

Since we've checked subscription to the server template in the rule above here we only need check create_uid.

Aldeigja commented:
> Consider moving out some common methods into `setUp`. Also you can remove some intermediate operations by optimising the test flow.

I think it would be better to create a new PR for test refactoring.

Aldeigja commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,425 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        self.server_test_1.message_unsubscribe([self.user_bob.partner_id.id])
+        with self.assertRaises(AccessError):
+            plan_line_action_read_result = plan_line_action_as_bob.read([])
+
```
> **Extract common test patterns into shared utilities.**
> 
> The test method has significant overlap with `test_plan_lines_access_rights`. Consider refactoring to reduce duplication.

We implement this approach in test refactoring.

Aldeigja commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,413 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        self.server_test_1.message_unsubscribe([self.user_bob.partner_id.id])
+        with self.assertRaises(AccessError):
+            plan_line_read_result = plan_line_as_bob.read([])
+
```
> The test method lacks a docstring explaining the test cases.

Docstrings have been added.

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,503 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        self.add_to_group(self.user_bob, "cetmix_tower_server.group_manager")
+        test_plan_2_as_bob.write({"access_level": "2"})
+        self.assertEqual(test_plan_2_as_bob.access_level, "2")
+        test_plan_2_as_bob.line_ids.write({"sequence": 3})
```
Using the `sequence` field may affect the order of the lines inside the plan. Is this what we want to do?
If not, then we should use some other field, eg `path`.

Aldeigja replied:
I'm trying to emulate string reordering here.

Aldeigja replied:
Fixed

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,503 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        )
+
+        # Ensure the plan line was created and check that create_uid is user_bob
+        self.assertEqual(
```
This is handled by ORM, we cannot affect this, so no need to test it either.

Aldeigja replied:
Fixed

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,503 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+
+        # Add user_bob to `group_manager` and test plan.line.action access
+        self.add_to_group(self.user_bob, "cetmix_tower_server.group_manager")
+        # Set access level to 2, so group_manager can access the plan line action
```
Manager has access to records with the "User" access level. No need to change the access lever unless this is done for some reason.

Aldeigja replied:
Fixed

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,503 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+            )
+        )
+
+        self.assertEqual(
```
No need to check the ORM methods we don't modify.

Aldeigja replied:
Fixed

ivs-cetmix commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1261,3 +1261,503 @@ def test_plan_with_another_plan_with_all_not_executable_condition(self):
+        # User subscribed to the server: ensure user_bob cannot remove variable value
+        # from action line if he is not creator of action line
+        with self.assertRaises(AccessError):
+            test_plan_line_action_as_bob.write({"variable_value_ids": [(5, 0, 0)]})
```
This doesn't remove the variable value, but removes all variable values from the x2m field. 
Although in this case it's supposed to be the same, yet we need to use more explicit way of doing that.
Eg what if there are more than one value and error is raised because of those extra ones?

Aldeigja replied:
Fixed

ivs-cetmix reviewed (changes requested):
Looks good, please check comments though.

ivs-cetmix commented on cetmix_tower_server/security/cx_tower_variable_value_security.xml:
```diff
@@ -1,18 +1,121 @@
+        <field name="model_id" ref="model_cx_tower_variable_value" />
+        <field name="domain_force">
+            [
+            ('create_uid', '=', user.id)
```
This rule will allow Manager group users to delete any record they created. Even if they don't have read access to it any longer.

ivs-cetmix reviewed (approved):
Intermediate LGTM, need to check tests once again though.

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

GabbasovDinar reviewed (approved):
LGTM

Aldeigja commented on cetmix_tower_server/tests/test_plan.py:
```diff
@@ -1271,3 +1305,390 @@ def test_plan_unlink(self):
+            self.variable_value.with_user(self.user_bob).unlink()
+        # User subscribed to the server: ensure user_bob can create variable value in
+        # plan line action
+        variable_value_as_bob = (
+            self.env["cx.tower.variable.value"]
+            .with_user(self.user_bob)
+            .create(
+                {
+                    "variable_id": self.variable_os.id,
+                    "value_char": "OS/2",
+                }
+            )
+        )
```
> _⚠️ Potential issue_
> 
> **Fix the variable value creation test case.**
> 
> The variable value is created without a `plan_line_action_id`, which doesn't align with the test's context of testing variable values associated with plan line actions. Consider:
> 
> ```diff
> -                {
> -                    "variable_id": self.variable_os.id,
> -                    "value_char": "OS/2",
> -                }
> +                {
> +                    "variable_id": self.variable_os.id,
> +                    "value_char": "OS/2",
> +                    "plan_line_action_id": self.plan_3_line_1_action.id
> +                }
> ```
> 
> 📝 Committable suggestion

This variable value is assigned to plan line action created later with same user, please check string [#1676](https://github.com/cetmix/cetmix-tower/blob/31103401765bec52d661f4781571082b76e4b947/cetmix_tower_server/tests/test_plan.py#L1676)

ivs-cetmix commented:
/ocabot merge nobump

CetmixGitDrone commented:
Hey, thanks for contributing! Proceeding to merge this for you.
Prepared branch [14.0-dev-ocabot-merge-pr-98-by-ivs-cetmix-bump-nobump](https://github.com/cetmix/cetmix-tower/commits/14.0-dev-ocabot-merge-pr-98-by-ivs-cetmix-bump-nobump), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 3bb89e85f83e352b302d6b6c44b5550b1229bd9d. Thanks a lot for contributing to cetmix. ❤️
